### PR TITLE
Add `SYNC` command to internal ZooKeeper client

### PR DESCRIPTION
**Pull request by antonio2368**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Support for `SYNC` was added to the internal client.
Within the session, we already have a strict ordering of operations, requests are processed 1 by 1 as they are received. E.g. local reads are not done until the previous writes are finished. That means that with a SYNC (or any write operation) we can already block local reads until that write is finished.
Because all the writes go through Raft, we are guaranteed to have sequentially consistent operations by sending SYNC or a write operation on a reconnect.
I added the sync operation to our internal client so it's more clear what is our goal when we call it instead of using a dummy write operation.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Common/ZooKeeper/ZooKeeper.cpp [hunk 1]
```diff
@@ -1145,6 +1173,37 @@ Coordination::Error ZooKeeper::tryMultiNoThrow(const Coordination::Requests & re
...
+    auto callback = [promise](const Coordination::SyncResponse & response) mutable
+    {
+        promise->set_value(response);
+    };
+
+    impl->sync(path, std::move(callback));
+    return future;
+}
+
+std::future<Coordination::SyncResponse> ZooKeeper::asyncSync(const std::string & path)
```
Comment (tavplubix) [inline src/Common/ZooKeeper/ZooKeeper.cpp:1190]: This method is unused now (and probably we don't need it at all). But maybe it makes sense to have it for consistency with similar methods for other requests.

File: src/Common/ZooKeeper/ZooKeeperImpl.cpp [hunk 2]
```diff
@@ -1,3 +1,4 @@
+#include "Common/ZooKeeper/ZooKeeperCommon.h"
```
Comment (tavplubix) [inline src/Common/ZooKeeper/ZooKeeperImpl.cpp:1]:
```suggestion
#include <Common/ZooKeeper/ZooKeeperCommon.h>
```

**Review by tavplubix — APPROVED** [review]

**Comment by tavplubix** [comment]
Currently, it's kind of dead code that is not used and therefore not covered by tests. Could you please add some usage of `ZooKeeper::sync` btw? For example, here:
https://github.com/ClickHouse/ClickHouse/blob/98a89b50ffa0d1cfc14ace910bc6acd4d6287fe9/src/Interpreters/TransactionLog.cpp#L238

File: src/Common/ZooKeeper/ZooKeeper.cpp (hunk 1 above)
Comment (antonio2368) [reply src/Common/ZooKeeper/ZooKeeper.cpp:1190]: yes, the only reason was because of consistency.

**Comment by antonio2368** [comment]
@Mergifyio update

**Comment by antonio2368** [comment]
[ClickHouse Stateless Tests (release, Databasereplicated, Actions) [2/2]](https://s3.amazonaws.com/clickhouse-test-reports/38047/bdd5b94192bc5df841a4e2202c32fff0522eb5d7/stateless_tests__release__databasereplicated__actions__[2/2].html) - fails on `0167_strange_mutations`, seems unrelated because `sync` is used only for transaction logs (it passed the first time)
ClickHouse Integration Tests (release, Actions) [2/2] - OOM

**Comment by tavplubix** [comment]
> [ClickHouse Stateless Tests (release, Databasereplicated, Actions) [2/2]](https://s3.amazonaws.com/clickhouse-test-reports/38047/bdd5b94192bc5df841a4e2202c32fff0522eb5d7/stateless_tests__release__databasereplicated__actions__%5B2/2%5D.html) - fails on `0167_strange_mutations`, seems unrelated because `sync` is used only for transaction logs (it passed the first time) ClickHouse Integration Tests (release, Actions) [2/2] - OOM

SELECT failed with `Table test_yvjieh.test doesn't exist`, because table was not created, because of
```
022.06.21 10:04:24.272684 [ 100788 ] {7c7dc44c-7581-4fe9-972c-03ea4402918a} <Error> executeQuery: Code: 122. DB::Exception: Table columns structure in ZooKeeper is different from local table structure. Local columns:
columns format version: 1
2 columns:
`n` UInt64
`test` UInt8

Zookeeper columns:
columns format version: 1
1 columns:
`n` Int32
. (INCOMPATIBLE_COLUMNS) (version 22.7.1.345) (from 0.0.0.0:0) (comment: '01162_strange_mutations.sh') (in query: /* ddl_entry=query-0000000016 */ CREATE TABLE test_yvjieh.test UUID '52c17aef-e577-4c5e-9193-0e6926ce858e' (`n` UInt64, `test` UInt8) ENGINE = ReplicatedMergeTree('/test/01162/01162_strange_mutations_test_yvjieh/auto_{shard}', '1auto_{replica}') ORDER BY n AS SELECT number + 100 AS n, 0 AS test FROM numbers(50)), Stack trace (when copying this message, always include the lines below):
```
It might happen if table `t` was not not completely removed from zk on DROP. But DROP was executed synchronously:
```
2022.06.21 10:04:24.171588 [ 100787 ] {4c4d1a18-47c6-46f5-9709-59a463490396} <Debug> executeQuery: (from [::1]:45958) (comment: '01162_strange_mutations.sh') drop table t (stage: Complete)
2022.06.21 10:04:24.171623 [ 100787 ] {4c4d1a18-47c6-46f5-9709-59a463490396} <Trace> ContextAccess (default): Access granted: DROP TABLE ON test_yvjieh.t
2022.06.21 10:04:24.171636 [ 100787 ] {4c4d1a18-47c6-46f5-9709-59a463490396} <Debug> DatabaseReplicated (test_yvjieh): Proposing query: DROP TABLE t NO DELAY
2022.06.21 10:04:24.173464 [ 100787 ] {4c4d1a18-47c6-46f5-9709-59a463490396} <Debug> DDLWorker(test_yvjieh): Waiting for worker thread to process all entries before query-0000000015
2022.06.21 10:04:24.173610 [ 100787 ] {4c4d1a18-47c6-46f5-9709-59a463490396} <Debug> DDLWorker(test_yvjieh): Processing task query-0000000015 (DROP TABLE t NO DELAY)
2022.06.21 10:04:24.174147 [ 100787 ] {4c4d1a18-47c6-46f5-9709-59a463490396} <Debug> DDLWorker(test_yvjieh): Executing query: DROP TABLE test_yvjieh.t NO DELAY
2022.06.21 10:04:24.174222 [ 100787 ] {4c4d1a18-47c6-46f5-9709-59a463490396} <Debug> executeQuery: (from 0.0.0.0:0, user: ) (comment: '01162_strange_mutations.sh') /* ddl_entry=query-0000000015 */ DROP TABLE test_yvjieh.t NO DELAY (stage: Complete)
2022.06.21 10:04:24.174245 [ 100787 ] {4c4d1a18-47c6-46f5-9709-59a463490396} <Information> test_yvjieh.t (91707260-5efb-4ff8-8eca-192e6bbec7af): Stopped being leader
2022.06.21 10:04:24.174262 [ 100787 ] {4c4d1a18-47c6-46f5-9709-59a463490396} <Trace> test_yvjieh.t (ReplicatedMergeTreeRestartingThread): Restarting thread finished
2022.06.21 10:04:24.174743 [ 100787 ] {4c4d1a18-47c6-46f5-9709-59a463490396} <Trace> test_yvjieh.t (ReplicatedMergeTreeRestartingThread): Waiting for threads to finish
2022.06.21 10:04:24.174770 [ 100787 ] {4c4d1a18-47c6-46f5-9709-59a463490396} <Trace> test_yvjieh.t (ReplicatedMergeTreeRestartingThread): Threads finished
2022.06.21 10:04:24.174777 [ 100787 ] {4c4d1a18-47c6-46f5-9709-59a463490396} <Trace> test_yvjieh.t (PartMovesBetweenShardsOrchestrator): PartMovesBetweenShardsOrchestrator thread finished
2022.06.21 10:04:24.175621 [ 100787 ] {4c4d1a18-47c6-46f5-9709-59a463490396} <Debug> DatabaseCatalog: Waiting for table 91707260-5efb-4ff8-8eca-192e6bbec7af to be finally dropped
2022.06.21 10:04:24.182022 [ 100787 ] {4c4d1a18-47c6-46f5-9709-59a463490396} <Debug> DDLWorker(test_yvjieh): Executed query: DROP TABLE test_yvjieh.t NO DELAY
2022.06.21 10:04:24.183235 [ 100787 ] {4c4d1a18-47c6-46f5-9709-59a463490396} <Information> executeQuery: Read 3 rows, 141.00 B in 0.011623542 sec., 258 rows/sec., 11.85 KiB/sec.
2022.06.21 10:04:24.183261 [ 100787 ] {4c4d1a18-47c6-46f5-9709-59a463490396} <Debug> MemoryTracker: Peak memory usage (for query): 2.03 MiB.
``` 

So it looks like #34818

**Pull request merged**